core/drivers/scmi-msg/smt.c

```diff
@@ -0,0 +1,225 @@
+
+/* SMP protection on channel access */
+static unsigned int smt_channels_lock;
+
+/* If channel is not busy, set busy and return true, otherwise return false */
+static bool channel_set_busy(struct scmi_msg_channel *chan)
+{
+	bool channel_is_busy = false;
+	uint32_t exceptions = 0;
+
+	if (chan->threaded)
+		exceptions = thread_mask_exceptions(THREAD_EXCP_FOREIGN_INTR);
+
+	cpu_spin_lock(&smt_channels_lock);
+
+	channel_is_busy = chan->busy;
+	if (!channel_is_busy)
+		chan->busy = true;
+
+	cpu_spin_unlock(&smt_channels_lock);
+
+	if (chan->threaded)
+		thread_unmask_exceptions(exceptions);
+
+	return !channel_is_busy;
+}
+
+static void channel_release_busy(struct scmi_msg_channel *chan)
+{
+	chan->busy = false;
+}
+
+static struct smt_header *channel_to_smt_hdr(struct scmi_msg_channel *chan)
+{
+	return (struct smt_header *)io_pa_or_va(&chan->shm_addr);
+}
+
+/*
+ * Creates a SCMI message instance in secure memory and push it in the SCMI
+ * message drivers. Message structure contains SCMI protocol meta-data and
+ * references to input payload in secure memory and output message buffer
+ * in shared memory.
+ */
+static void scmi_proccess_smt(unsigned int agent_id, uint32_t *payload_buf)
+{
+	struct scmi_msg_channel *chan = NULL;
+	struct smt_header *smt_hdr = NULL;
+	size_t in_payload_size = 0;
+	uint32_t smt_status = 0;
+	struct scmi_msg msg = { };
+	bool error = true;
+
+	chan = plat_scmi_get_channel(agent_id);
+	if (!chan)
+		return;
+
+	smt_hdr = channel_to_smt_hdr(chan);
+	assert(smt_hdr);
+
+	smt_status = atomic_load_u32(&smt_hdr->status);
+
+	if (!channel_set_busy(chan)) {
+		DMSG("SCMI channel %u busy", agent_id);
+		goto out;
+	}
+
+	in_payload_size = atomic_load_u32(&smt_hdr->length) -
+			  sizeof(smt_hdr->message_header);
+
+	if (in_payload_size > SCMI_PLAYLOAD_MAX) {
+		DMSG("SCMI payload too big %u", in_payload_size);
+		goto out;
+	}
+
+	if (smt_status & (SMT_STATUS_ERROR | SMT_STATUS_FREE)) {
+		DMSG("SCMI channel bad status 0x%x",
+		     smt_hdr->status & (SMT_STATUS_ERROR | SMT_STATUS_FREE));
+		goto out;
+	}
+
+	/* Fill message */
+	msg.in = (char *)payload_buf;
+	msg.in_size = in_payload_size;
+	msg.out = (char *)smt_hdr->payload;
+	msg.out_size = chan->shm_size - sizeof(*smt_hdr);
+
+	assert(msg.out && msg.out_size >= sizeof(int32_t));
+
+	/* Here the payload is copied in secure memory */
+	memcpy(msg.in, smt_hdr->payload, in_payload_size);
+
+	msg.protocol_id = SMT_HDR_PROT_ID(smt_hdr->message_header);
+	msg.message_id = SMT_HDR_MSG_ID(smt_hdr->message_header);
+	msg.agent_id = agent_id;
+
+	scmi_process_message(&msg);
+
+	/* Update message length with the length of the response message */
+	smt_hdr->length = msg.out_size_out + sizeof(smt_hdr->message_header);
+
+	channel_release_busy(chan);
+	error = false;
+
+out:
+	if (error) {
+		DMSG("SCMI error");
+		smt_status |= SMT_STATUS_ERROR;
+	}
+
+	smt_status |= SMT_STATUS_FREE;
+
+	atomic_store_u32(&smt_hdr->status, smt_status);
```

Comment:
This is written to without an explicit lock. How is that supposed to work? If it's OK please add a short comment explaining why this is OK.

Reply:
ok, I'll add a inline comment.

Reply:
checking back I made it this way to ensure `status` in shared memory is update only once, not in 2 steps, in case compiler updates it for `STATUS_ERROR` then for `STATUS_FREE`. But I seems to me now that this is useless.
For clarity, i'll remove the atomic store and change to:
```c
	if (error) {
		DMSG("SCMI error");
		smt_hdr->status |= SMT_STATUS_ERROR | SMT_STATUS_FREE;
	} else {
		smt_hdr->status |= SMT_STATUS_FREE;
	}